File: Solution/994.腐烂的橘子/994.腐烂的橘子.py
'''
Descroption: LeetCode 994. 腐烂的橘子
Author: EmoryHuang
Date: 2021-08-04 15:58:17
Method:
bfs
首先将所有腐烂的橘子加入队列
bfs查找它的相邻元素，若为新鲜橘子，则入队，每查找一层，时间加一
最后遍历矩阵，若还有新鲜橘子，则返回-1
'''


class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        dx = [0, -1, 0, 1]
        dy = [1, 0, -1, 0]
        mins = 0
        q = deque()
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 2:
                    q.append([i, j, mins])
        while q:
            lenq = len(q)
            for _ in range(lenq):
                x, y, mins = q.popleft()
                for i in range(4):
                    new_x = x + dx[i]
                    new_y = y + dy[i]
                    if 0 <= new_x < n and 0 <= new_y < m and grid[new_x][
                            new_y] == 1:
                        grid[new_x][new_y] = 2
                        q.append([new_x, new_y, mins + 1])
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    return -1
        return mins

File: Solution/994.腐烂的橘子/994.腐烂的橘子.java
/*
 * @Descroption: LeetCode 994. 腐烂的橘子
 * @Author: EmoryHuang
 * @Date: 2023-07-19 09:54:05
 * @Method:
 * BFS
 * 从每个腐烂的橘子开始
 */

import java.util.*;

class Solution {
    public int orangesRotting(int[][] grid) {
        int[][] dirs = { { -1, 0 }, { 0, 1 }, { 1, 0 }, { 0, -1 } };
        int m = grid.length, n = grid[0].length;
        int total = 0;
        Queue<int[]> queue = new LinkedList<>();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 2) {
                    queue.offer(new int[] { i, j });
                }
                total += grid[i][j] == 0 ? 0 : 1;
            }
        }
        int ans = 0, cnt = 0;
        while (!queue.isEmpty()) {
            int len = queue.size();
            cnt += len;
            boolean flag = false;
            for (int i = 0; i < len; i++) {
                int[] cur = queue.poll();
                int x = cur[0], y = cur[1];
                for (int j = 0; j < 4; j++) {
                    int nx = x + dirs[j][0], ny = y + dirs[j][1];
                    if (nx >= 0 && nx < m && ny >= 0 && ny < n && grid[nx][ny] == 1) {
                        grid[nx][ny] = 2;
                        queue.offer(new int[] { nx, ny });
                        flag = true;
                    }
                }
            }
            ans += flag ? 1 : 0;
        }
        return total == cnt ? ans : -1;
    }
}